### PR TITLE
fix(eslint-plugin): [no-unnec-type-assert] handle JSX attributes

### LINKED ISSUE #982: [no-unnecessary-type-assertion] doesn't handle types properly -> React['key']
**Repro**

`no-unnecessary-type-assertion` is resolving `key` keys on React components  inconsistently with normal function call. Please take a look at the code for better descrition.

 

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  }
}
```

```ts
const appTest = () => {
  type ArtificialKey = string | number;
  interface ReactAttributeArtificial {
    key?: ArtificialKey;
  }
  type Data = {
    id?: null | string | number;
  };

  const Test = (props: Data) => {
    // ✅no tsc errors
    // ✅no eslint errors
    return ChildFn({ key: props.id! });
  };

  const Test2 = (props: Data) => {
    // 🚨ESLINT error.
    // 🚨If you remove non null operator it will be an compile error
    return <ChildCmp key={props.id!} />;
  };

  const ChildFn = (props: ReactAttributeArtificial) => 'Hello';
  const ChildCmp = () => <div>Hello</div>;
};
```

**Expected Result**

It should behave consistently

**Actual Result**

JSX `key` is resolved in not correct way ?

image

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.2.0` |
| `@typescript-eslint/parser`        | `2.2.0` |
| `TypeScript`                       | `3.6.3` |
| `ESLint`                           | `6.3.0` |
| `node`                             | `10` |
| `yarn`                              | `1.17` |

### LINKED ISSUE #1061: [no-unnecessary-type-assertion] False positive with null/undefined union type
Looks like the fix for #529 did not cover all edge cases.

```JSON
{
  "rules": {
      "@typescript-eslint/no-unnecessary-type-assertion": "error",
  }
}
```

```TS
const optionalNonNull = (s?: string) => { };
let s: string | null | undefined;

/*
Argument of type 'string | null | undefined' is not assignable to parameter of type 'string | undefined'.
  Type 'null' is not assignable to type 'string | undefined'.
*/
optionalNonNull(s);

/*
This assertion is unnecessary since the receiver accepts the original type of the expression.
*/
optionalNonNull(s!);
```

**Expected Result**

Should not trigger an error.

**Actual Result**

Triggers an error.

**Additional Info**

N/A

**Versions**

```
"@typescript-eslint/eslint-plugin": "^2.3.3",
```

Also tested with master 054df278d6b7064a44b5f78fec453bf9ae6ad281

### PR DESCRIPTION
**Pull request by bradzacher**
Fixes #982
Fixes #1061

I never added proper handling for JSX attributes in the rule - this fixes that.

This also fixes a few things with testing TSX files
- the rule tester code (from the 2.0 changes) didn't allow you to change the filename
- there was no tsx test file
- there was weird behaviour from the tests if I didn't clean the caches between tests
    - it wasn't actually testing the tsx code for some reason.

### CODE REVIEW

**Comment by armano2** [comment]
@bradzacher i pushed merge + all conflicts are resolved 
(most of changes are no longer necessary as of comments was completely changed)

File: packages/eslint-plugin/package.json [hunk 1]
```diff
@@ -50,6 +50,7 @@
     "@types/json-schema": "^7.0.3",
     "@types/marked": "^0.7.1",
     "@types/prettier": "^1.18.2",
+    "@types/react": "^16.9.2",
```
Comment (armano2) [inline packages/eslint-plugin/package.json]: do we really have to import react types? i don't see them being necessary to run this test

File: packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts [hunk 2]
```diff
@@ -117,6 +124,19 @@ function testFunction(_param: string | null): void { /* noop */ }
...
+      code: `
+const React = require("react");
+
+function Test(props: {
+  id?: null | string | number;
+}) {
+  return <div key={props.id!} />;
+};
+      `,
+      filename: path.join(rootDir, 'react.tsx'),
```
Comment (armano2) [inline packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts]: filename for now has to be absolute: #1424 should fix this

File: packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts [hunk 3]
```diff
@@ -14,7 +14,7 @@ const ruleTester = new RuleTester({
 });
 
 // make sure each test is completely isolated
-// there was some weird behaviour with the mixed ts/tsx test cases without this
+// there was some weird behavior with the mixed ts/tsx test cases without this
```
Comment (armano2) [inline packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts]: https://grammarist.com/spelling/behavior-behaviour/

File: packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts [hunk 4]
```diff
@@ -12,6 +13,12 @@ const ruleTester = new RuleTester({
   parser: '@typescript-eslint/parser',
 });
 
+// make sure each test is completely isolated
+// there was some weird behavior with the mixed ts/tsx test cases without this
+afterEach(() => {
+  clearCaches();
```
Comment (armano2) [inline packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts]: this is actually no longer needed

File: packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts [hunk 5]
```diff
@@ -117,6 +124,19 @@ function testFunction(_param: string | null): void { /* noop */ }
 const value = 'test' as string | null | undefined
 testFunction(value!)
     `,
+    // https://github.com/typescript-eslint/typescript-eslint/issues/982
+    {
+      code: `
+const React = require("react");
```
Comment (armano2) [inline packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts]:
```suggestion
declare namespace JSX {
    interface IntrinsicElements {
        div: { key: string | number }
    }
}
```

File: packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts [hunk 6]
```diff
@@ -327,5 +347,33 @@ class Mx {
         },
       ],
     },
+    // https://github.com/typescript-eslint/typescript-eslint/issues/982
+    {
+      code: `
+const React = require("react");
```
Comment (armano2) [inline packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts]:
```suggestion
declare namespace JSX {
    interface IntrinsicElements {
        div: { key?: string | number }
    }
}
```

File: packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts (hunk 4 above)
Comment (armano2) [inline packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts]: this is actually no longer needed

File: packages/eslint-plugin/package.json (hunk 1 above)
Comment (armano2) [inline packages/eslint-plugin/package.json]: instead importing types its better to use (in tests):
```ts
declare namespace JSX {
    interface IntrinsicElements {
        div: { key?: string | number }
    }
}
```

**Comment by armano2** [comment]
@bradzacher i picked up this PR as it looked that you abandoned it, 

i pushed merge + all conflicts are resolved and cleanup up code a little (most of changes was outdated and already supported on master)

i left some comments ^Up to show what i changed

**Comment by armano2** [comment]
note: all changes related to #1061 are already present on master 
https://github.com/typescript-eslint/typescript-eslint/pull/1002/commits/df410a11963a19f052b13c2e1f6dadf429ae7bc2#diff-9c31441197b8f835651744ab117e4b6dR223-R232

File: packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts (hunk 3 above)
Comment (bradzacher) [reply packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts]: Yeah, I write Australian english (which is pretty much the same as UK english). Hence I put a `u` in behaviour, colour and favourite.

Which is why I hated setting the spellchecker to american english.. haha

File: packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts (hunk 2 above)
Comment (bradzacher) [reply packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts]: do we want to put this in the `RuleTester`?

**Comment by bradzacher** [comment]
Thanks, this was on my todo list. I had some commits locally, but never got around to finishing fixing it up.

File: packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts (hunk 2 above)
Comment (armano2) [reply packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts]: no it should not be there, it should work correctly with relative paths
see #1439

**Pull request merged**